examples/voice-pipeline-agent/multi_task/news_mailer.py

```diff
@@ -0,0 +1,204 @@
+    email = user_data.get("email")
+    news = user_data.get("news")
+
+    if not email:
+        return "email is not collected"
+
+    if not news:
+        return "news is not collected"
+
+    # mock sending email
+    query = user_data.get("query")
+    logger.info(f"Sending news about {query} to {email}")
+    await asyncio.sleep(2)
+    return f"The news about {query} is sent to {email}"
+
+
+@llm.ai_callable()
+async def verify_email(
+    email: Annotated[str, llm.TypeInfo(description="The collected email address")],
+) -> str:
+    """Called to verify the user's email address."""
+    if "@" not in email:
+        return "The email address is not valid, please confirm with the user."
+
+    # Potentially show the email on the screen for the user to confirm
+    return "The email address is valid. Please confirm with the user for the spelling."
+
+
+@llm.ai_callable()
+async def update_email(
+    email: Annotated[str, llm.TypeInfo(description="The collected email address")],
+) -> str:
+    """Called to update the user's email address."""
+
+    agent = AgentCallContext.get_current().agent
+    user_data: UserData = agent.user_data
+    user_data["email"] = email
+    logger.info(f"The email is updated to {email}")
+    return f"The email is updated to {email}."
+
+
+news_mailer = AgentTask(
+    name="news_mailer",
+    instructions=(
+        "You are a friendly assistant that can query news from the internet."
+        "Summarize the news in 50 words or less and ask the user if they want to receive the news by email."
+        "Use email_collector to collect the user's email address."
+    ),
+    functions=[query_news, send_news_email],
+)
+
+email_collector = AgentTask(
+    name="email_collector",
+    instructions=(
+        "You are a friendly assistant that can collect the user's email address. Your tasks:\n"
+        "1. Collect the user's email address, help to complete the @ and domain part if possible.\n"
+        "2. Verify the address with `verify_email` function until the user confirms.\n"
+        "3. Update the email address after the user confirms.\n"
+        "Transfer back to news_mailer after the email is updated."
+    ),
+    functions=[update_email, verify_email],
+)
+
+
+async def entrypoint(ctx: JobContext):
+    await ctx.connect(auto_subscribe=AutoSubscribe.AUDIO_ONLY)
+
+    chat_log_file = "news_mailer.log"
+
+    # Set up chat logger
+    chat_logger = logging.getLogger("chat_logger")
+    chat_logger.setLevel(logging.INFO)
+    handler = logging.FileHandler(chat_log_file)
+    formatter = logging.Formatter("%(message)s")
+    handler.setFormatter(formatter)
+    chat_logger.addHandler(handler)
+
+    participant = await ctx.wait_for_participant()
+    agent = VoicePipelineAgent(
+        vad=ctx.proc.userdata["vad"],
+        stt=deepgram.STT(),
+        llm=openai.LLM(),
+        tts=openai.TTS(),
+        initial_task=news_mailer,
+        available_tasks=[news_mailer, email_collector],
```

Comment:
how does it know which tasks to use in this case? is it purely from the instructions indicating `Use email_collector to collect the user's email address.`.. does that mean `email_collector` is represented as a function?

Reply:
each task has a built in `transfer_to` method that will be added to `fnc_ctx` if the `can_enter` of the task returns true.